volpiano_display_utilities/text_volpiano_alignment.py

```diff
@@ -6,238 +6,317 @@
-    # Squash final syllables of a word together if there are more
-    # syllables in the text than in the volpiano.
-    if len(text_syls) > len(vol_syls):
-        txt_syls_wo_overhang, txt_overhanging_syls = (
-            text_syls[: len(vol_syls) - 1],
-            text_syls[len(vol_syls) - 1 :],
-        )
-        txt_joined_overhanging_syls = "".join(txt_overhanging_syls)
-        text_syls = txt_syls_wo_overhang + [txt_joined_overhanging_syls]
-        # Add spacing to volpiano to account for combining the
-        # overhanging text syllables together into the final
-        # syllable.
-        vol_syls[-1] += "-" * (len(txt_joined_overhanging_syls) - len(vol_syls[-1]) + 1)
-    # Pad text syllables with empty strings if more syllables in the
-    # volpiano than in the text.
-    comb_wrd: List[Tuple[str, str]] = list(
-        zip_longest(text_syls, vol_syls, fillvalue="")
-    )
-    return comb_wrd
+    zipped_word = _zip_and_align(text, volpiano, pad_text="", pad_volpiano="---")
+    aligned_word = []
+    for text_syl, vol_syl in zipped_word[:-1]:
+        vol_syl = adjust_music_spacing(vol_syl, len(text_syl), False)
+        aligned_word.append((text_syl, vol_syl))
+    text_syl, vol_syl = zipped_word[-1]
+    vol_syl = adjust_music_spacing(vol_syl, len(text_syl), True)
+    aligned_word.append((text_syl, vol_syl))
+    return aligned_word
 
 
 def _align_section(
-    text_section: List[List[str]], volpiano_section: str
+    text_section: SyllabifiedTextSection,
+    volpiano_section: SyllabifiedVolpianoSection,
 ) -> List[Tuple[str, str]]:
     """
-    Aligns a section of text and volpiano, padding in case
-    either text or volpiano is longer or shorter for that section.
+    Aligns a section of text and volpiano. In sections of syllabified
+    text, these are aligned word by word, padding (with additional words:
+    "----" in the case of volpiano and "" in the case of text) if
+    either the two sections have different numbers of words. In sections
+    of unsyllabified text, the text is aligned to the entire volpiano
+    section.
 
-    text_section [list[list[str]]]: nested list of syllabized text. Each
-        sublist represents a word, and each element of the sublist represents
-        a syllable (or unsyllabified phrase where text is not to be syllabified).
-    volpiano_section [str]: volpiano string for the section
+    text_section [SyllabifiedTextSection]: section of text
+    volpiano_section [SyllabifiedVolpianoSection]: section of volpiano
 
-    returns [list[tuple[str, str]]]: list of tuples of text syllables and
-        volpiano syllables
+    returns [list[tuple[str, str]]]: list of tuples of text and volpiano syllables
     """
-    logging.debug("Aligning section - Text: %s", text_section)
-    logging.debug("Aligning section - Volpiano: %s", volpiano_section)
+    logging.debug(
+        "Aligning section - Text: %s. Volpiano: %s", text_section, volpiano_section
+    )
```

Comment:
not necessary to change, but you could use `logging.debug(f"Aligning section - {text_section=}, {volpiano_section=}.")` to achieve a similar result, and it should all fit on one line.

Reply:
Interesting! Thanks.